File: src/main/java/com/lmmmowi/leetcode/p393/Solution.java
package com.lmmmowi.leetcode.p393;

/**
 * @Author: mowi
 * @Date: 2022/4/14
 * @Description: 393. UTF-8 编码验证[https://leetcode-cn.com/problems/utf-8-validation/]
 */
public class Solution {

    private static final int BYTE_10000000 = 0x80;
    private static final int BYTE_11000000 = 0xC0;

    public boolean validUtf8(int[] data) {
        int n = data.length;
        int ptr = 0;
        while (ptr < n) {
            int firstByte = data[ptr++];
            if ((firstByte & BYTE_10000000) == 0) {
                continue;
            }

            int len = detectBytesLength(firstByte);
            if (len <= 1 || len > 4) {
                return false;
            }

            for (int i = 0; i < len - 1; i++) {
                if (ptr >= n || (data[ptr++] & BYTE_11000000) != BYTE_10000000) {
                    return false;
                }
            }
        }
        return true;
    }

    private int detectBytesLength(int b) {
        int len = 0;
        while ((b & BYTE_10000000) > 0) {
            len++;
            b <<= 1;
        }
        return len;
    }
}

File: src/main/go/com/lmmmowi/leetcode/go/p393/Solution.go
// 393. UTF-8 编码验证[https://leetcode-cn.com/problems/utf-8-validation/]
package main

const byte10000000 int = 0x80
const byte11000000 int = 0xC0

func validUtf8(data []int) bool {
	n := len(data)
	ptr := 0

	for ptr < n {
		firstByte := data[ptr]
		ptr++
		if (firstByte & byte10000000) == 0 {
			continue
		}

		length := detectBytesLength(firstByte)
		if length <= 1 || length > 4 {
			return false
		}

		for i := 0; i < length-1; i++ {
			if ptr >= n || (data[ptr]&byte11000000) != byte10000000 {
				return false
			}
			ptr++
		}
	}

	return true
}

func detectBytesLength(b int) int {
	length := 0
	for (b & byte10000000) > 0 {
		length++
		b <<= 1
	}
	return length
}
